### PR TITLE
Support non literal position and length for substring

### PR DESCRIPTION
**Pull request by firestarman**
closes https://github.com/NVIDIA/spark-rapids/issues/6887

Support non literal position and length for substring.

cudf `substring(column, column, column)`  behaves differently to `substring(column, int, int)` when `start` or `end` is negative, so methods calling them took care of the differences.

***Performance***
- About 5.2 GB Parquet data in local files.
- CPU 12 cores, and one GPU (Titan V, with 12GB memory)

| CPU | GPU |
| ------------ | ------------ | 
| 8.057 s  | 5.269 s  |

Signed-off-by: Firestarman <firestarmanllc@gmail.com>

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by revans2 — COMMENTED** [review]
I really would like to see some performance numbers around using the JIT. If it is really a lot better, then we are going to need to make sure that we do some documentation/configuration/testing so we are sure no user will run into problems, especially with the JIT cache. We should also pay attention to first run vs later runs and cached vs not cached.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -460,34 +461,169 @@ case class GpuSubstring(str: Expression, pos: Expression, len: Expression)
...
+      // need to do the similar calculation as CPU here.
+      poses.add(strSizes, DType.INT32)
+    }
+    withResource(negConvertedPoses) { _ =>
+      val negPosFlags = withResource(Scalar.fromInt(0)) { zero =>
+        poses.lessThan(zero)
+      }
+      withResource(negPosFlags) { _ =>
+        // 2) (pos > 0) ? pos -1 : 0
+        val subOne = "__device__ inline void f(int* out,int pos){*out=(pos>0)?(pos-1):pos;}"
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:479]: This is not for perf when I chose to use JIT here, just want to keep code as simple as possible.
Good to know its disadvantages, will update it.

**Comment by firestarman** [comment]
Remove "databricks" from the title since DB has been verified in previous runs.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 2]
```diff
@@ -460,34 +461,169 @@ case class GpuSubstring(str: Expression, pos: Expression, len: Expression)
...
-      val0: GpuColumnVector,
-      val1: GpuColumnVector,
-      val2: GpuColumnVector): ColumnVector =
-        throw new UnsupportedOperationException(s"Cannot columnar evaluate expression: $this")
+  private[this] def computeStarts(strs: ColumnView, poses: ColumnView): ColumnVector = {
+    // CPU:
+    //     start = (pos < 0) ? pos + str_size : ((pos > 0) ? pos - 1 : 0)
+    // 1) pos + str_size
+    val negConvertedPoses = withResource(strs.getCharLengths) { strSizes =>
+      // cudf `substring(column, column, column)` treats negative start always as 0, so
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**